File: pathToSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

/* 113. Path Sum II
Time Complexity: O(n)
Space Complexity:O(n)
*/
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(root, sum, result, path);
        return result;
    }

    private void dfs(TreeNode root, int sum, List<List<Integer>> result, List<Integer> path){

        //Case 1: root is null
        if(root == null)
            return;

        path.add(root.val);
        if(root.left == null && root.right == null){
            if(sum == root.val)
                result.add(new ArrayList<>(path));
        }

        //Case 3: left is null
        if(root.right != null){
            dfs(root.right, sum-root.val, result, path);
            path.remove(path.size() - 1);
        }

        //Case 4: right is null
        if(root.left != null){
            dfs(root.left, sum-root.val, result, path);
            path.remove(path.size() - 1);
        }
    }
}


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<List<Integer>> result;
    int sum1 = 0;

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        result = new ArrayList<>();
        int sum_to_leaf = 0;
        sum1 = sum;

        List<Integer> path = new ArrayList<>();
        helper(root, path, sum_to_leaf);
        return result;
    }

    private void helper(TreeNode root, List<Integer> path, int sum_to_leaf){
        if(root == null)
            return;

        path.add(root.val);
        sum_to_leaf += root.val;

        if(root.right == null && root.left == null){
            if(sum_to_leaf == sum1)
                result.add(path);
        }

        helper(root.right, new ArrayList<>(path), sum_to_leaf);
        helper(root.left, new ArrayList<>(path), sum_to_leaf);
        return;
    }
}

File: pathToSum.py
"""
Intuition: Store the path till the node in each node
#####################################################################
Time Complexity : O(N) 
Space Complexity : O(H)
#####################################################################
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> List[List[int]]:
        
        result = []
        self.pathSumHelper(root, [], 0, result, targetSum)
        return result
    
    def pathSumHelper(self, node, partialPath, partialSum, result, targetSum):
        
        if node is None:
            return
        
        totalPath = partialPath + [node.val]
        totalSum = partialSum + node.val
        
        if node.left is None and node.right is None:
            if targetSum == totalSum:
                result.append(totalPath)
        self.pathSumHelper(node.left, totalPath, totalSum, result, targetSum)
        self.pathSumHelper(node.right, totalPath, totalSum, result, targetSum)
            